[Kotlin Warnings As Errors] Card Reader Module - Resolve Warnings & Enable All Warnings as Errors

ParaskP7 opened the pull request:
<!-- Remember about a good descriptive title. -->

Parent: #6816
Closes: #6818

<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR resolves a couple of warnings for the `cardreader` module and then enables all warnings as errors on it as well.

Warnings Resolution List:
1. [Resolve global scope delicate coroutines api warning.](https://github.com/woocommerce/woocommerce-android/commit/f64ccbacbdcc5ecfc9f6bd35e54f074cc366ecb0)
2. [Resolve on failure named arguments function calling problem.](https://github.com/woocommerce/woocommerce-android/commit/b926d89c5c4792466adeb548804a4fab3862330b)

The `allWarningsAsErrors` configuration is currently applied on the module level in order to make sure that, as the overall [Kotlin Warnings as Errors](https://github.com/woocommerce/woocommerce-android/issues/6816) work is progressing, no new warnings are added to this module, which is already free of warnings.

When the overall [Kotlin Warnings as Errors](https://github.com/woocommerce/woocommerce-android/issues/6816) work is complete on all modules, then this module level `allWarningsAsErrors` configuration will be replaced by a root level `allWarningsAsErrors` configuration that will be applied by default to all modules from there after (see [here](https://github.com/woocommerce/woocommerce-android/issues/6820)).

-----

PS: @wzieba I added you as the main reviewer, that is, in addition to the @woocommerce/owl-team team itself, but randomly, since I just wanted someone from the `Woo Android` team to sing-off on that change for WCAndroid as well. PS: I am going to randomly add more of you in those PRs just so you become more aware of this change and how close we are on enabling `allWarningsAsErrors` by default everywhere. 🥇 

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

- There is nothing much to test here.
- Verifying that all the CI checks are successful should be enough.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

ParaskP7 committed:
Analysis: Resolve global scope delicate coroutines api warning

1st Warning Message: "This is a delicate API and its use requires care.
Make sure you fully read and understand documentation of the declaration
that is marked as a delicate API."

Adding the '@OptIn(DelicateCoroutinesApi::class)' annotation for the
'cancelPayment(...)' function is the recommended action to resolve this
'GlobalScope' warnings. However, this results to the below '@OptIn'
warnings:

2nd Warning Message: "This annotation should be used with the
compiler argument '-Xopt-in=kotlin.RequiresOptIn'"

Adding the '-Xopt-in=kotlin.RequiresOptIn' free compiler argument is the
recommended action to resolve this '@OptIn' warning.

ParaskP7 committed:
Analysis: Resolve on failure named arguments function calling problem

Warning Message: "The corresponding parameter in the supertype
'PaymentIntentCallback' is named 'e'. This may cause problems when
calling this function with named arguments"

Rename 'exception' to 'e' is the recommended action to resolve this
'onFailure(...)' warnings.

ParaskP7 committed:
Analysis: Enable all warnings as errors for card reader module

This 'allWarningsAsErrors' configuration is currently applied on the
module level in order to make sure that, as the overall 'Kotlin Warnings
as Errors' work is progressing, no new warnings are added to this
module, which is already free of warnings.

When the overall 'Kotlin Warnings as Errors' work is complete on all
modules, then this module level 'allWarningsAsErrors' configuration will
be replaced by a root level 'allWarningsAsErrors' configuration that
will be applied by default to all modules from there after.

ParaskP7 commented:
FYI: I am converting this to draft since I need to fix the unit test related warnings for this module as well (see [here](https://buildkite.com/automattic/woocommerce-android/builds/4102#0181aaa2-d3d8-4531-9fe7-be26f11c7537)). 

Command: `./gradlew lib:cardreader:testDebug`

ParaskP7 committed:
Analysis: Resolve variable dummy reader id is never used test warning

Warning Message: "Variable 'dummyReaderId' is never used"

Deleting these unused 'dummyReaderId' variables is the recommended
action to resolve this type of warnings.

ParaskP7 committed:
Analysis: Resolve run blocking test is deprecated test warning

Warning Message: "'runBlockingTest' is deprecated. Use 'runTest' instead
to support completing from other dispatchers. Please use the migration
guide for details: https://github.com/Kotlin/kotlinx.coroutines/blob/
master/kotlinx-coroutines-test/MIGRATION.md"

Replacing 'runBlockingTest' with 'testBlocking' from
'CardReaderBaseUnitTest' is the recommended action to resolve this type
of warnings, as per the general direction for the test on this repo and
more specifically for this module (since the 'WooCommerce' module
utilizes the 'BaseUnitTest' instead, which is pretty much identical).

ParaskP7 committed:
Analysis: Resolve test function name characters test problem

Warning Message: "Name contains characters which can cause problems on
Windows: "

Renaming '""' to 'empty' resolves this 'name character' warning.

ParaskP7 commented:
👋 @wzieba !

Are you able to pick this up this week or are you busy and would prefer if I assign this PR to someone else?

wzieba commented:
hi @ParaskP7 👋 , sorry for the delay - yes, I'll review this PR this week. Thanks for the ping!

ParaskP7 commented:
Hey, thanks @wzieba and I am in no much rush, I just wanted to make sure of it is progressing as I would expect, and in case you are not available, for me to lift this PR off your shoulders. 💯 

PS: Spoiler alert, its an easy one... famous last words! 😅 🤞

wzieba commented on libs/cardreader/build.gradle:
```diff
@@ -25,8 +25,12 @@ android {
+        }
         freeCompilerArgs += [
-                "-Xopt-in=kotlinx.coroutines.ExperimentalCoroutinesApi"
+                "-Xopt-in=kotlinx.coroutines.ExperimentalCoroutinesApi",
```
I know it was like this before, but as we have this compiler arg, can we in scope of this PR remove all `@ExperimentalCoroutinesApi` annotations in the `cardreader` module?

ParaskP7 replied:
👋 @wzieba !

1. I am not sure what you mean by removing all the `@ExperimentalCoroutinesApi` annotations in the `cardreader` module. I am only seeing those being used in tests and rightfully so. What I can do as part of this PR is improve the `@ExperimentalCoroutinesApi` on tests and apply them only on the class level or remove those when and if not necessary. Wdyt?
2. Maybe instead we should remove this `"-Xopt-in=kotlinx.coroutines.ExperimentalCoroutinesApi"` free compiler argument instead. I don't think we need it, right?

ParaskP7 replied:
Oh, actually you're right, there is an unnecessary `@ExperimentalCoroutinesApi` annotation on the main source related `DiscoverReadersAction` class that I believe can be removed. 👍 Let me look at that... 🤔

ParaskP7 replied:
👋 @wzieba !

1. I did the change here: 4b6f4ad4d05e96435ccdc4bd58a991f90df716ee
2. I also added these extra test related changes: d7f9336ed1e6504e2e943ec4452cc9a9069f53e6 + d3e2e9d0bec5662243399adbb26e54ac0da1cc51

Let me know what you think. 🙇

wzieba replied:
Oh, I didn't mean to remove the compiler arg. I think we want to have it, it was deliberately added in 6f606a035c3660a362fd34c60619da0395e14bfc.

What I meant is that, as we have this compiler arg, we don't need to have `@ExperimentalCoroutinesApi`. As this annotation is obsolete, I suggested removing it.

wzieba replied:
So I think the change `DiscoverReadersAction` was the one I was thinking about.

>  I am only seeing those being used in tests and rightfully so.

Could you please share more context on why the annotation should be used in tests? I feel like I'm missing something.

ParaskP7 replied:
Yea, I got that, but I do think that we can and should remove this `compiler arg` anyway. 😅  I can share more details on why we might want to remove that and be explicit with the `@ExperimentalCoroutinesApi` annotations from now on and when needed, that is, instead of having them enabled by default like that. Wdyt?

Also, as you can see, me removing this `compiler arg` already helped with tests as we had to explicitly add the annotation there, just because of that, otherwise the test compilation would was failing. Obviously, the same will apply for the `main` source code.

PS.1: Adding @malinajirka to the discussion since he added that. Maybe he can share more insights about why he did that then. 🤔 
PS.2: This `compiler arg` is only added to this module and I haven't seen that used in any other module, or client (like `WPAndroid`) or library (like `FluxC`). Maybe, for us to be consistent everywhere, it is better if we don't have this added to this module too. 🤔

ParaskP7 replied:
> Could you please share more context on why the annotation should be used in tests? I feel like I'm missing something.

Now that the `compiler arg` is removed, you can try and remove this `@ExperimentalCoroutinesApi` annotation and see what happens, this will give you a hint on why it is needed. It is just that with this `compiler arg` the need for the `@ExperimentalCoroutinesApi` annotation got hidden, and thus I vote for it being removed anyway (see my comment above).

wzieba replied:
> I can share more details on why we might want to remove that and be explicit with the @ExperimentalCoroutinesApi annotations from now on and when needed, that is, instead of having them enabled by default like that. Wdyt?

Yes, that would be great! 

> Now that the compiler arg is removed, you can try and remove this @ExperimentalCoroutinesApi annotation and see what happens, this will give you a hint on why it is needed.

<img width="592" alt="image" src="https://user-images.githubusercontent.com/5845095/177815819-48b0a7f5-2e91-4c37-95b0-207b8b098049.png">

Do you mean this explanation? If yes, it didn't really convince me 😅 .

> Maybe, for us to be consistent everywhere, it is better if we don't have this added to this module too. 🤔

I don't see benefits on being consistent on that matter here!

---

IMO, this annotation is redundant. I don't believe that someone writing code in Woo acts differently if they see an "experimental coroutines API" warning. We use it in the whole codebase for years now, this API is not anything new. I think it just adds a really not needed alt+enter meaningless work. So I'd vote to keep the compiler arg, just remove all `@ExperimentalCoroutinesApi` from this module and whole codebase as I think it doesn't bring any value.

WDYT?

ParaskP7 replied:
> Do you mean this explanation? If yes, it didn't really convince me 😅 .

Well, it is not that this will convince you, or that this is the explanation, this is rather the warning that you need to look into and click into it to go deeper in order to really see which of the APIs is experimental and why.

> IMO, this annotation is redundant. I don't believe that someone writing code in Woo acts differently if they see an "experimental coroutines API" warning. We use it in the whole codebase for years now, this API is not anything new. I think it just adds a really not needed alt+enter meaningless work. So I'd vote to keep the compiler arg, just remove all @ExperimentalCoroutinesApi from this module and whole codebase as I think it doesn't bring any value.

I see your point and I understand where you coming from, it feels off to have and needing this `@ExperimentalCoroutinesApi` annotation everywhere, also having it propagating from one class to another, and not getting any value from it... I feel you! 💯 

However, I have a different opinion on those, and find value in them, especially since with every `Kotlin` and/or `Coroutine` update we get a warning to remove those when they are no longer needed. Also, now that we have a better annotation, the `@OptIn(ExperimentalCoroutinesApi::class)`, which at least help us avoid propagating the annotation everywhere, it feels that we have a good balance and could have those targeted on exactly where they are needed.

But, this is just my opinion. I think it might be better if we create a P2 post about it and ask the engineers how they feel about it, get some more opinions and then decide on whether we want to suppress the `@ExperimentalCoroutinesApi` everywhere, for all modules, for the whole project. Wdyt?

ParaskP7 replied:
No matter, for the scope of this PR, I think we are getting a bit out of hand here, wdyt? 😅 

This was not a warning anyway and we could have merged this PR without any work on it, right?

wzieba replied:
Yes, I agree 😅  - it's not a blocker and we could've merged this PR without any work. I just thought it was a good opportunity to discuss it, it's not a burning issue by any means!

ParaskP7 replied:
As per our internal discussion, I'll proceed with merging this PR and we will deal with adding back the `-Xopt-in=kotlinx.coroutines.ExperimentalCoroutinesApi` free compiler argument later on and as a separate topic. Once again, this was a nice discussion and thank you for opening it up @wzieba ! ❤️ 🥇 💯

wzieba reviewed (approved):
Looks good to me, leaving one comment 👍 Thanks for taking care of it!

ParaskP7 commented:
Thank you for the review and for your comment @wzieba ! 🙇 🥇

ParaskP7 committed:
Analysis: Remove unnecessary experimental coroutines api annotation

The '@ExperimentalCoroutinesApi' annotation can be now removed from the
'DiscoverReadersAction' class. As such, this
'-Xopt-in=kotlinx.coroutines.ExperimentalCoroutinesApi' configuration
can be also removed altogether from 'freeCompilerArgs'. The below
'-Xopt-in=kotlin.RequiresOptIn' configuration is all that is needed from
now on.

ParaskP7 committed:
Analysis: Move experimental coroutines api annotation to class level

This is done because:
1) Of having one '@ExperimentalCoroutinesApi' on the test class level
instead of multiple, one for each test function is simpler.
2) Of consistency purposes as all other test suites are using the
'@ExperimentalCoroutinesApi' annotation.

ParaskP7 committed:
Analysis: Add missing experimental coroutines api annotation for tests
